### PR TITLE
Add `Basis.derivatives`

### PR DESCRIPTION
**Pull request by inducer**
cc @a-alveyblanc

### CODE REVIEW

**Comment by inducer** [comment]
It also simplifies the tensor product derivative spaghetti by quite a bit.

File: modepy/test/test_modes.py [hunk 1]
```diff
@@ -198,6 +199,13 @@ def test_basis_grad(dim, shape_cls, order, basis_getter):
             logger.info("error: %.5e", err)
             eoc_rec.add_data_point(h, err)
 
+            for idiff_axis in range(shape.dim):
+                dr_bf = basis.functions_dr(idiff_axis)[ifunc](r)
+                ref = gradbf_v[idiff_axis]
+                err = la.norm(dr_bf - ref, 2)
+                if la.norm(ref, 2) > 1e-13:
+                    err = err/la.norm(ref, 2)
```
Comment (inducer) [inline modepy/test/test_modes.py:207]: Whoops. This isn't testing anything. :facepalm: On it.

Comment (inducer) [reply modepy/test/test_modes.py:207]: Now actually tested

File: modepy/modes.py [hunk 2]
```diff
@@ -766,6 +791,28 @@ def gradients(self) -> tuple[BasisGradientType, ...]:
         Each function returns a tuple of derivatives, one per reference axis.
         """
 
+    def functions_dr(self, axis: int) -> tuple[BasisFunctionType, ...]:
```
Comment (inducer) [inline modepy/modes.py:794]: Still unsure about the name. Suggestions?

Comment (inducer) [reply modepy/modes.py:794]: Maybe just `derivatives`? `deriv_functions`?

Comment (inducer) [reply modepy/modes.py]: I agree, that's probably the sanest. Renamed to that in the latest.

File: modepy/matrices.py [hunk 3]
```diff
@@ -131,6 +131,12 @@ def multi_vandermonde(
     A sequence of matrices is returned--i.e. this function
     works directly on :func:`modepy.Basis.gradients` and returns
     a tuple of matrices.
+
+    .. note::
+
+        Depending on the use case, it may be convenient to instead call
+        :func:`vandermonde` with the result of :meth:`Basis.derivatives`,
+        which is mathematically equivalent.
```
Comment (alexfikl) [inline modepy/matrices.py:139]: Maybe phrase this more like "it also works with Basis.derivatives"? It's a bit unclear what use cases could / should work.

**Review by alexfikl — APPROVED** [review]
Left a little documentation nitpick, but otherwise this looks good to me!

File: modepy/modes.py (hunk 2 above)
Comment (alexfikl) [reply modepy/modes.py]: `axis_derivative_functions`? 

I'm also leaning towards `derivatives`, so that's perfectly fine by me!

Comment (inducer) [reply modepy/modes.py]: `axis_derivatives` is another option, but it's unlikely that we'd have any other type (directional?) derivative, so :shrug:. I say let's leave it as `derivatives`.

File: modepy/matrices.py (hunk 3 above)
Comment (inducer) [reply modepy/matrices.py]: Did some rephrasing, though I'm not sure I covered what you asked for here. This is in the docstring for `multi_vandermonde`, which won't work with `Basis.derivatives`. I did specify the use case more clearly. LMK if this works.

Comment (alexfikl) [reply modepy/matrices.py]: Yeah, that's a lot clearer to me. Thanks!

**Pull request merged**